### PR TITLE
Make all logic implemented inside PSP22Data struct

### PR DESCRIPTION
**Pull request by h4nsu**
Moves all PSP22 token logic from the example implementation (`lib.rs`) inside the `PSP22Data` struct. Thanks to that the logic is now fixed and a contract developer can use our implementation by simply implementing `PSP22` trait in a "pass-through" way:
```
impl PSP22 for MyContract {
    ...
    #[ink(message)]
    fn balance_of(&self, owner: AccountId) -> u128 {
        self.data.balance_of(owner)
    }
    ...
    #[ink(message)]
    fn transfer(&mut self, to: AccountId, value: u128, data: Vec<u8>) -> Result<(), PSP22Error> {
        let events = self.data.transfer(self.env().caller(), to, value, data)?;
        self.emit_events(events); 
        Ok(())
    }
    ...
}
```

This approach:
a) greatly reduces the amount of boilerplate code that needs to be copied into user's contract
b) avoids using complicated macro on top of `ink` macro
c) (not sure if this is good, but OpenBrush had it) gives the user control over emitting events

### CODE REVIEW

File: lib.rs [hunk 1]
```diff
@@ -24,6 +25,23 @@ mod token {
                 data: PSP22Data::new(supply, Self::env().caller()),
             }
         }
+
+        fn emit_events(&self, events: ink::prelude::vec::Vec<PSP22Event>) {
+            for event in events {
+                match event {
+                    PSP22Event::Transfer(from, to, value) => {
+                        self.env().emit_event(Transfer { from, to, value })
```
Comment (pmikolajczyk41) [inline lib.rs:33]: not sure if `Transfer` type is available outside contract module, but if so, you might want to implement conversion (`Into`) from `PSP22Event` type to this 'native' event type

the advantage is that we encapsulate event type matching close to the event definition, and outside, like here, we can just go `for e in events { self.env().emit_event(e.into()) }`

but this might be infeasible (due to opaque event types) and is pretty much personal taste, so feel free to ignore this note comment :)

File: lib.rs [hunk 2]
```diff
@@ -38,29 +56,26 @@ mod token {
     #[ink(event)]
     pub struct Transfer {
         #[ink(topic)]
-        from: AccountId,
+        from: Option<AccountId>,
         #[ink(topic)]
-        to: AccountId,
+        to: Option<AccountId>,
```
Comment (pmikolajczyk41) [inline lib.rs:61]: when `to` can be `None`?

File: lib.rs [hunk 3]
```diff
@@ -24,6 +25,23 @@ mod token {
                 data: PSP22Data::new(supply, Self::env().caller()),
             }
         }
+
+        fn emit_events(&self, events: ink::prelude::vec::Vec<PSP22Event>) {
```
Comment (pmikolajczyk41) [inline lib.rs:29]: sad thing is that this method has to be copied for every psp22 contract... maybe we could extract it from here and expose somehow for reuse?

if you go with the other suggestion, you might end up with a very generic, convenient method for emitting arbitrary event iterable from a contract, which sounds nice :)

**Review by pmikolajczyk41 — APPROVED** [review]
Delegation seems to be clean solution (although might be a bit onerous for rich traits), so approving with a few minor comments.

There is one hidden thing that we'd better be careful about (especially when adapting this approach for other PSPs). The default implementation on the `PSP22Data` type should meet the assumption that no storage change is done if the method returns `Err(_)`. A situation when this might lead to a big problem is such a custom delegation:
```rust
fn method(args) {
    if let Err(Event::MinorProblem) = self.data.method(args) {
        self.do_some_custom_fallback()
    }
}
```
In other words, we can't assume that a failed 'template' method will lead to a failed call. As far as I can see, your code meets this requirement, so we are ok.

File: lib.rs (hunk 1 above)
Comment (h4nsu) [reply lib.rs:33]: Unfortunately events are not visible outside the scope of `ink::contract` macro... yet. AFAIK ink! 5 will introduce a possibility to define events externally. After that happens, the whole enum->event conversion will be removed and `PSP22Data` methods will simply return vectors of native events. Before that happens, automating the conversion would require putting some very cryptic boilerplate inside the contract. Since this is meant to be simple educational example, I'd rather go with stupid&simple approach.

File: data.rs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    Approval(AccountId, AccountId, u128),
+}
+
+#[ink::storage_item]
+#[derive(Debug, Default)]
+pub struct PSP22Data {
+    total_supply: u128,
+    balances: Mapping<AccountId, u128>,
+    allowances: Mapping<(AccountId, AccountId), u128>,
+}
```
Comment (deuszx) [inline data.rs:19]: It'd be good to have comments (`///`) on pub elements. Especially this is supposed to be an exemplary usage.

File: data.rs [hunk 5]
```diff
@@ -0,0 +1,170 @@
+use crate::PSP22Error;
+use ink::{
+    prelude::{vec, vec::Vec},
+    primitives::AccountId,
+    storage::Mapping,
+};
+
+pub enum PSP22Event {
+    Transfer(Option<AccountId>, Option<AccountId>, u128),
```
Comment (deuszx) [inline data.rs:9]:
```suggestion
    Transfer { 
        from: Option<AccountId>, 
        to: Option<AccountId>, 
        amount: u128
        },
```
This is a user-facing type that they're supposed to map into a proper `PSP22Event` in their contracts. We don't want them to mess up the fields.

File: data.rs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    Approval(AccountId, AccountId, u128),
```
Comment (deuszx) [inline data.rs:10]: Similarly here.

File: lib.rs (hunk 2 above)
Comment (deuszx) [reply lib.rs:61]: Or `from`. It doesn't seem to be representable in the API of PSP22 standard - i.e. you can't transfer from/to `None`.

File: data.rs [hunk 7]
```diff
@@ -0,0 +1,170 @@
+use crate::PSP22Error;
```
Comment (deuszx) [inline data.rs:1]: We might think about adding a mod-level documentation here, explaining why _this_ representation and how it's supposed to be integrated into other contracts.

File: lib.rs (hunk 2 above)
Comment (h4nsu) [reply lib.rs:61]: It's reserved for handling minting and burning tokens. I'm just trying to follow the standard defined here:
https://github.com/w3f/PSPs/blob/master/PSPs/psp-22.md#transfer

File: lib.rs (hunk 3 above)
Comment (h4nsu) [reply lib.rs:29]: Unfortunately it's probably not possible to extract a method which uses `self.env()` outside the contract. Or maybe technically it's possible, but results in some templating nightmare.

This method is here temporary, after ink! allows defining events outside contracts, it will be gone and all `self.emit_events(events)` will be replaces with `for e in events { self.env().emit_event(e); }`

Comment (deuszx) [reply lib.rs:29]: I think you _could_ extract (or rather implement in the `PSP22Data`) something along these lines:
```rust
 fn emit_event<C: ContractBase, EE: EmitEvent<C>>(emitter: EE, event: Event) {
            emitter.emit_event(event);
        }
```

File: data.rs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+pub enum PSP22Event {
+    Transfer(Option<AccountId>, Option<AccountId>, u128),
+    Approval(AccountId, AccountId, u128),
+}
+
+#[ink::storage_item]
+#[derive(Debug, Default)]
+pub struct PSP22Data {
+    total_supply: u128,
```
Comment (pmikolajczyk41) [inline data.rs:16]: maybe not in this PR, but I think we should be generic over balances type

File: data.rs [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        self.balances.get(owner).unwrap_or_default()
+    }
+
+    pub fn allowance(&self, owner: AccountId, spender: AccountId) -> u128 {
+        self.allowances.get((owner, spender)).unwrap_or_default()
+    }
+
+    pub fn transfer(
+        &mut self,
+        caller: AccountId,
```
Comment (deuszx) [inline data.rs:46]:
```suggestion
        from: AccountId,
```

At this level of abstraction there's no notion of "caller" anymore.

File: data.rs (hunk 4 above)
Comment (h4nsu) [reply data.rs:19]: I'll add nice comments and extensive readme in a different PR

File: data.rs (hunk 8 above)
Comment (h4nsu) [reply data.rs:16]: Again, the standard specifies it to be strictly `u128`.

Comment (h4nsu) [reply data.rs:16]: Well, to be precise, the standard defines a type alias `type Balance = u128;` which introduces name collision with the ink env type for native chain currency. Hence I decided to drop that alias.

File: data.rs (hunk 5 above)
Comment (h4nsu) [reply data.rs]: Good idea, done

File: data.rs (hunk 6 above)
Comment (h4nsu) [reply data.rs]: Done

File: data.rs (hunk 9 above)
Comment (h4nsu) [reply data.rs:54]: These internal methods of `PSP22Data` differ in signature form the ones defined by the `PSP22` trait. My idea was to make it explicit, that this argument is meant to always be `self.env().caller()` when someone is implementing the trait.

File: data.rs (hunk 7 above)
Comment (h4nsu) [reply data.rs:1]: Will be added soon (tm) ;)

**Review by deuszx — APPROVED** [review]
